drivers/mem_manage/mm_drv_common.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Common Memory Management Driver Code
+ *
+ * This file provides common implementation of memory management driver
+ * functions, for example, sys_mm_drv_map_region() can use
+ * sys_mm_drv_map_page() to map page by page for the whole region.
+ * This avoids duplicate implementations of same functionality in
+ * different drivers. The implementations here are marked as
+ * weak functions so they can be overridden by the driver.
+ */
+
+#include <kernel.h>
+#include <string.h>
+#include <toolchain.h>
+#include <sys/check.h>
+#include <sys/util.h>
+
+#include <drivers/mem_manage/system_mem_manage.h>
+
+#include "mm_drv_common.h"
+
+struct k_spinlock sys_mm_drv_common_lock;
+
+bool sys_mm_drv_is_addr_array_aligned(uintptr_t *addr, size_t cnt)
+{
+	size_t idx;
+	bool ret = true;
+
+	for (idx = 0; idx < cnt; idx++) {
+		if (!sys_mm_drv_is_addr_aligned(addr[idx])) {
+			ret = false;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+bool sys_mm_drv_is_virt_region_mapped(void *virt, size_t size)
+{
+	size_t offset;
+	bool ret = true;
+
+	for (offset = 0; offset < size; offset += CONFIG_MM_DRV_PAGE_SIZE) {
+		uint8_t *va = (uint8_t *)virt + offset;
+
+		if (sys_mm_drv_page_phys_get(va, NULL) != 0) {
+			ret = false;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+bool sys_mm_drv_is_virt_region_unmapped(void *virt, size_t size)
+{
+	size_t offset;
+	bool ret = true;
+
+	for (offset = 0; offset < size; offset += CONFIG_MM_DRV_PAGE_SIZE) {
+		uint8_t *va = (uint8_t *)virt + offset;
+
+		if (sys_mm_drv_page_phys_get(va, NULL) != -EFAULT) {
+			ret = false;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+int sys_mm_drv_simple_map_region(void *virt, uintptr_t phys,
+				 size_t size, uint32_t flags)
+{
+	k_spinlock_key_t key;
+	int ret = 0;
+	size_t offset;
+
+	CHECKIF(!sys_mm_drv_is_addr_aligned(phys) ||
+		!sys_mm_drv_is_virt_addr_aligned(virt) ||
+		!sys_mm_drv_is_size_aligned(size)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	key = k_spin_lock(&sys_mm_drv_common_lock);
+
+	for (offset = 0; offset < size; offset += CONFIG_MM_DRV_PAGE_SIZE) {
+		uint8_t *va = (uint8_t *)virt + offset;
+		uintptr_t pa = phys + offset;
+
+		int ret2 = sys_mm_drv_map_page(va, pa, flags);
+
+		if (ret2 != 0) {
+			ret = ret2;
```

Comment:
in case we fail to map one page we keep going and we map all the remaining and return the error. But now I guess the user cannot use that page at that `va` and he doesn't even know which page failed to be mapped. How is this supposed to be used?

Reply:
I thought of using `CHECKIF()` but decided against it as you can disable the check with kconfig, but I forgot to add back the assertion.

As for what you have stated, I don't have a good solution as we cannot simply revert changes made before the failure because there is no history saved. Returning a failed list would probably require a few malloc() which I would like to avoid. I could force a kernel panic here if you want. Note that this provides a very primitive implementation where an app specific solution can override this.

Reply:
You could keep history on the stack with a recursive call here, but that's probably evil as you don't know the stack size of the calling thread and are likely to run it over very quickly.